azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/core/legacy/LegacyRefreshTokenCredentialFactory.java

```diff
@@ -0,0 +1,128 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+public class LegacyRefreshTokenCredentialFactory {
+    public static Mono<TokenCredential> createRefreshTokenCredential(AzureEnvironment environment, TokenCredential tokenCredential) {
+        return tokenCredential.getToken(new TokenRequestContext()
+                .addScopes(ScopeUtil.resourceToScopes(environment.managementEndpoint())))
+                .map(accessToken -> {
+                    IAuthenticationResult result = ((MsalToken) accessToken).getAuthenticationResult();
+                    String refreshTokenFromResult;
+                    try {
+                        refreshTokenFromResult = (String) FieldUtils.readField(result, "refreshToken", true);
+                    } catch (IllegalAccessException e) {
+                        throw new AzureToolkitRuntimeException("Cannot read refreshToken from IAuthenticationResult.");
+                    }
+                    if (StringUtils.isBlank(refreshTokenFromResult)) {
+                        throw new AzureToolkitRuntimeException("Fail to get refresh token.");
+                    }
+
+                    List<String> tenantIds = listTenantIds(environment, tokenCredential);
+                    if (CollectionUtils.isEmpty(tenantIds)) {
+                        throw new AzureToolkitRuntimeException("There are no tenants in your account.");
+                    }
+
+                    return LegacyRefreshTokenCredentialFactory.fromRefreshToken(
+                            environment, IdentityConstants.DEVELOPER_SINGLE_SIGN_ON_ID, tenantIds.get(0), refreshTokenFromResult);
+                });
+    }
+
+    public static TokenCredential fromRefreshToken(AzureEnvironment env, String clientId, String tenantId, String refreshToken) {
+        return buildRefreshTokenCredential(env, clientId, tenantId, refreshToken);
+    }
+
+    private static TokenCredential buildRefreshTokenCredential(AzureEnvironment env, String clientId, String tenantId, String refreshToken) {
+        return request -> Mono.fromCallable(() -> {
+            AuthenticationResult result =
+                    authorize(env,
+                            clientId,
+                            StringUtils.firstNonBlank(tenantId, "common"),
+                            refreshToken,
+                            StringUtils.isBlank(tenantId) ? null : ScopeUtil.scopesToResource(request.getScopes())
+                    );
+            return fromAuthenticationResult(result);
+        });
+    }
+
+    private static AccessToken fromAuthenticationResult(AuthenticationResult authenticationResult) {
+        if (authenticationResult == null) {
+            return null;
+        }
+        OffsetDateTime expiresOnDate = authenticationResult.getExpiresOnDate() == null ? OffsetDateTime.MAX :
```

Comment:
`MAX`? access token should be short-lived, I don't think the access token should last such a long time.
Azure may have a document about this, otherwise check OAuth2's  suggestion or some good practices.